sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/CertificateValidationCheck.cs

```diff
@@ -0,0 +1,310 @@
+            var parentBlock = variable.FirstAncestorOrSelf<BlockSyntax>();
+            if (parentBlock != null)
+            {
+                allAssignedExpressions.AddRange(parentBlock.DescendantNodes().OfType<AssignmentExpressionSyntax>()
+                    .Where(x => x.Left is IdentifierNameSyntax Ident && Ident.Identifier.ValueText == variable.Identifier.ValueText)
+                    .Select(x => x.Right));
+            }
+            if (variable.Initializer != null)       //Declarator initializer is counted as (default) assignment as well
+            {
+                allAssignedExpressions.Add(variable.Initializer.Value);
+            }
+            return MultiExpressionSublocations(c, allAssignedExpressions);
+        }
+
+        private static ImmutableArray<Location> InvocationLocations(InspectionContext c, MethodDeclarationSyntax method)
+        {
+            var returnExpressionSublocationsList = method.Body.DescendantNodes()
+                .OfType<ReturnStatementSyntax>()
+                .Select(x => x.Expression)
+                .Where(x => !IsVisited(c, x));      //Ignore all return statements with recursive call. Result depends on returns that could return compilant validator.
+            return MultiExpressionSublocations(c, returnExpressionSublocationsList);
+        }
+
+        private static bool IsVisited(InspectionContext c, ExpressionSyntax expression)
+        {
+            if (expression is InvocationExpressionSyntax invocation)
+            {
+                var MS = c.Context.SemanticModel.GetSymbolInfo(invocation).Symbol;
+                return MS != null && MS.DeclaringSyntaxReferences.Length != 0 && MS.DeclaringSyntaxReferences.Select(x => x.GetSyntax() as MethodDeclarationSyntax).Any(x => x != null && c.VisitedMethods.Contains(x));
+            }
+            return false;
+        }
+
+        private static ImmutableArray<Location> MultiExpressionSublocations(InspectionContext c, IEnumerable<ExpressionSyntax> expressions)
+        {
+            var exprSublocationsList = expressions.Distinct(new Helpers.CSharp.CSharpSyntaxNodeEqualityComparer<ExpressionSyntax>())
+                .Select(x => CallStackSublocations(c, x))
+                .ToArray();
+            if (exprSublocationsList.Any(x => x.Length == 0))   //If there's at leat one concurrent expression, that returns compilant delegate, than there's some logic and this scope is compliant
+            {
+                return ImmutableArray<Location>.Empty;
+            }
+            return exprSublocationsList.SelectMany(x => x).ToImmutableArray();      //Else every return statement is noncompliant
+        }
+
+        private static ImmutableArray<Location> CallStackSublocations(InspectionContext c, ExpressionSyntax expression)
+        {
+            var lst = ArgumentLocations(c, expression);
+            if (lst.Length != 0)        //There's noncompilant issue in this chain
```

Comment:
_minor_: we tend to use `!xxx.IsEmpty` syntax for readibility instead of `xxx.Length != 0`